### PR TITLE
Using of inheritance and polymorphism

### PR DESCRIPTION
**Pull request by TimurKM**
Please, review

### CODE REVIEW

File: TaskWithInheritanceAndPolymorphism/TaskWithInheritanceAndPolymorphism.cpp [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+	Figure()
+	{
+		std::cout << "Figure()" << std::endl;
+	}
+	~Figure()
+	{
+		std::cout << "~Figure()" << std::endl;
+	}
+
+	virtual const void what() = 0;
```
Comment (dimagerasimov) [inline TaskWithInheritanceAndPolymorphism/TaskWithInheritanceAndPolymorphism.cpp:17]: "const void" seriously? Please move the const specifier to the end of the function declaration like this:
`virtual void what() const = 0;`

The function will become a const function which means you can call it for constant objects as well.

 Please apply this comment to other functions.

File: TaskWithInheritanceAndPolymorphism/TaskWithInheritanceAndPolymorphism.cpp [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+		return area;
+	}
+	double const perimeter() override
+	{
+		double perimeter = 4 * m_length;
+		std::cout << perimeter << std::endl;
+		return perimeter;
+	}
+private:
+	double m_length;
```
Comment (dimagerasimov) [inline TaskWithInheritanceAndPolymorphism/TaskWithInheritanceAndPolymorphism.cpp:52]: Please declare the field like const double m_length. Use a member initializer list in the constructor. Your constructor will be: `Square(const double l) : l(length) { ...}`

File: TaskWithInheritanceAndPolymorphism/TaskWithInheritanceAndPolymorphism.cpp [hunk 3]
```diff
@@ -0,0 +1,139 @@
+﻿#include <iostream>
+#include <memory>
+#include <vector>
+
+class Figure
```
Comment (dimagerasimov) [inline TaskWithInheritanceAndPolymorphism/TaskWithInheritanceAndPolymorphism.cpp:5]: Please declare a Figure as a template class like:
`template <typename T>
class Figure`

Read how to forbid instantiating of this class using non-numeric types.

File: TaskWithInheritanceAndPolymorphism/TaskWithInheritanceAndPolymorphism.cpp (hunk 1 above)
Comment (TimurKM) [reply TaskWithInheritanceAndPolymorphism/TaskWithInheritanceAndPolymorphism.cpp]: Fixed

File: TaskWithInheritanceAndPolymorphism/TaskWithInheritanceAndPolymorphism.cpp (hunk 2 above)
Comment (TimurKM) [reply TaskWithInheritanceAndPolymorphism/TaskWithInheritanceAndPolymorphism.cpp]: Fixed

File: TaskWithInheritanceAndPolymorphism/TaskWithInheritanceAndPolymorphism.cpp (hunk 3 above)
Comment (TimurKM) [reply TaskWithInheritanceAndPolymorphism/TaskWithInheritanceAndPolymorphism.cpp]: Done with another way

File: TaskWithInheritanceAndPolymorphism/TaskWithInheritanceAndPolymorphism.cpp [hunk 4]
```diff
@@ -0,0 +1,135 @@
+﻿#include <iostream>
+#include <memory>
+#include <vector>
+
+template <typename T> class Figure
```
Comment (dimagerasimov) [inline TaskWithInheritanceAndPolymorphism/TaskWithInheritanceAndPolymorphism.cpp:5]: I don't see that you use T somewhere.

File: TaskWithInheritanceAndPolymorphism/TaskWithInheritanceAndPolymorphism.cpp [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+	{
+		std::cout << "~Figure()" << std::endl;
+	}
+
+	virtual void what() const = 0;
+	virtual double area() const = 0;
+	virtual double perimeter() const = 0;
+};
+
+class Square final : public Figure<double>
```
Comment (dimagerasimov) [inline TaskWithInheritanceAndPolymorphism/TaskWithInheritanceAndPolymorphism.cpp:22]: The Square and others should be a template as well as the Figure

File: TaskWithInheritanceAndPolymorphism/TaskWithInheritanceAndPolymorphism.cpp (hunk 4 above)
Comment (TimurKM) [reply TaskWithInheritanceAndPolymorphism/TaskWithInheritanceAndPolymorphism.cpp]: Used

File: TaskWithInheritanceAndPolymorphism/TaskWithInheritanceAndPolymorphism.cpp (hunk 5 above)
Comment (TimurKM) [reply TaskWithInheritanceAndPolymorphism/TaskWithInheritanceAndPolymorphism.cpp]: Used

File: TaskWithInheritanceAndPolymorphism/TaskWithInheritanceAndPolymorphism.cpp [hunk 6]
```diff
@@ -0,0 +1,139 @@
+﻿#include <iostream>
+#include <memory>
+#include <vector>
+
+template <typename T>
```
Comment (dimagerasimov) [inline TaskWithInheritanceAndPolymorphism/TaskWithInheritanceAndPolymorphism.cpp:5]: Please read and implement https://stackoverflow.com/questions/14294267/class-template-for-numeric-types. I like the static_assert solution.

Comment (TimurKM) [reply TaskWithInheritanceAndPolymorphism/TaskWithInheritanceAndPolymorphism.cpp:5]: Done

File: TaskWithInheritanceAndPolymorphism/TaskWithInheritanceAndPolymorphism.cpp [hunk 7]
```diff
@@ -49,6 +51,7 @@ class Square final : public Figure<T>
 		std::cout << perimeter << std::endl;
 		return perimeter;
 	}
+	static_assert(std::is_arithmetic<T>::value, "Template parameter must be numeric");
```
Comment (dimagerasimov) [inline TaskWithInheritanceAndPolymorphism/TaskWithInheritanceAndPolymorphism.cpp:54]: Isn't it enough to add static_assert only to the parent?

File: TaskWithInheritanceAndPolymorphism/TaskWithInheritanceAndPolymorphism.cpp [hunk 8]
```diff
@@ -1,7 +1,7 @@
 ﻿#include <iostream>
 #include <memory>
 #include <vector>
-
+#include <type_traits>
```
Comment (dimagerasimov) [inline TaskWithInheritanceAndPolymorphism/TaskWithInheritanceAndPolymorphism.cpp:4]: Add an empty line.

**Review by dimagerasimov — APPROVED** [review]
Great job!

**Pull request merged**